addons/keh_network/snapshotdata.gd

```diff
@@ -40,6 +40,8 @@ var _entity_name: Dictionary = {}
 
 # Holds the history of snapshots
 var _history: Array = []
+var _lifetime_history: Array = []
```

Comment:
Again, ideally the replay system should become its own addon. Still, there are a few things that can be considered here. The replay object can internally store all data as arrays of Snapshots, very much like the expected `_lifetime_history`. There is no need to have the encoded buffer as class object as it would be needed only to store the data in a file.

Also, the code style I would prefer to be more consistent with the rest of the addons, in which the static typing is very explicit, no like the line bellow using `:=` syntax.

Reply:
Decoupling replay functionality into a separate addon is a much better idea. I'll work up another prototype (and probably include a demo) to decouple the recording of replays. Will probably shift how the Replay class works from being a namespace with static functions for storing and reading serialized Arrays, to a replay with a history and its own `encdecbuffer`. Come to think of it, replays also need to store some metadata, too.

The only concerns that I have with entirely decoupling Replays from the Network addon is that currently replay data is sent by "intercepting" snapshots as they're being passed through the `_add_to_history` function that gets called from `_on_snapshot_finished`. Not immediately sure on how best to add a system that grabs the snapshot getting passed through the network singleton's `_update_control` without modifying the network script itself. For instance, if `replay.gd` becomes some kind of singleton below the network singleton, it could, for instance, swap out `_update_control`'s `sfinished` funcref to an alternative one that calls `network.on_snapshot_finished` and adds the snapshot to the replay's lifetime history before or after the call. But that seems like a really weird "spaghetti" way of tacking on replay functionality. But ultimately I want storing replays to be as unintrusive for someone to implement and performant as possible. Ideally someone would just "press a button" and whenever they're in a network session, they're also recording a replay. Of course, what with how replays are currently recorded, this doesn't even work for clients-- only the server/host records replays. My point is, I'll have a tricky time with making an easily activated replay system. Very open to suggestions on "where" in the network loop (if in the network loop at all) snapshots should be recorded.

In regards to syntax changes I can (and should) totally update those! It definitely makes things much clearer. I remember just wanting to prototype code quickly, so I skipped over some of the writing out of type definitions. (I'll also clean up some of the unhelpful comments).

Thanks for all the help!